### ISSUE TITLE
[QUIC] Design of Shutdown for QuicStreams in System.Net.Quic

### ISSUE THREAD

**Issue by jkotalik** [dotnet/runtime#756]
### Problem

So today, we have an API for Quic which looks like the following:
```c#
    public sealed class QuicStream : System.IO.Stream
    {
        internal QuicStream() { }
        public override bool CanSeek => throw null;
        public override long Length => throw null;
        public override long Seek(long offset, System.IO.SeekOrigin origin) => throw null;
        public override void SetLength(long value) => throw null;
        public override long Position { get => throw null; set => throw null; }
        public override bool CanRead => throw null;
        public override bool CanWrite => throw null;
        public override void Flush() => throw null;
        public override int Read(byte[] buffer, int offset, int count) => throw null;
        public override void Write(byte[] buffer, int offset, int count) => throw null;
        public long StreamId => throw null;
        public void AbortRead() => throw null;
        public ValueTask ShutdownWriteAsync(System.Threading.CancellationToken cancellationToken = default) => throw null;
    }
```

Today, QuicStream derives from Stream, which is really nice for interop purposes with other things, allowing people to just read and write to and from the stream without thinking what the underlying implementation is. 

When calling Dispose on the stream, that currently triggers an abortive shutdown sequence on the stream rather than graceful. To trigger a graceful shutdown, one needs to call ShutdownWriteAsync, which sends a FIN to the peer and waits for an ACK. However, people that would normally be using a stream wouldn't have access to ShutdownWriteAsync, hence all actions would be abortive by default. 

### Current thoughts

Ideally for the workflow we'd like to enable is something like:
```c#
using (Stream stream = connection.AcceptAsync()) // this is actually a quic stream
{
    Memory<byte> memory = new Memory<byte>(new byte[10]);
    int res = await stream.ReadAsync(memory);
    await stream.WriteAsync(memory);
}
```

The big question here is how do we make it so we can send a shutdown for a final write with a stream API outside of Dispose?

The first pivot would be whether we'd like QuicStream to derive from stream at all. If we'd like it to derive from System.IO.Stream, I think it would have to make Dispose/DisposeAsync do a graceful shutdown by default in that case, and expose Abortive versions of these functions on the QuicStream itself. That way, in the default case, using the stream APIs "just works". However, doing a graceful shutdown of a stream in Dispose has downsides. We would be doing a network call in Dispose, which would block. DisposeAsync also can't be canceled as well.

So let's talk about a world where QuicStream didn't derive from System.IO.Stream and we had a few liberties with the API. Potentially, we could change a few things:
- Make WriteAsync take a flag which says if this is the final write.
- Allow Dispose to be abortive.

Abortive Dispose wouldn't block at all, which is nice. However, we do need to be careful about how to supply an error code when aborting a stream. The error code that is used for abort is supposed to be "application specific and provided", meaning we'd need a default error code.

cc @dotnet/http3 and @nibanks I haven't really reached a conclusion based on this yet. I'd be interested in your thoughts.

**Comment by scalablecory** [comment]
Q: an abortive close also needs I/O, doesn't it? Is it managed behind the scenes, or does it also need to be async?

If we implement `Stream` -- and we probably should, because so many things can use it -- then `Dispose` and `DisposeAsync` should be graceful to match the expected behavior of `Stream`. We just need to live with that extra I/O, I think :/.

Lets look at `NetworkStream.Dispose`:
* It performs a graceful shutdown, so your sample code works great.
* It doesn't block, because of OS magic, unless you set linger options.
* If you're calling it during an exception-triggered unwind and your protocol doesn't have ways to distinguish a graceful shutdown from a busted protocol, you're going to have a bad time.
   * You need to set ownsSocket=false and manage shutdown yourself to get around this.

Spitballing... what we could do is enable an 'advanced' behavior like so:

```c#
using QuicStream stream = CreateUnidirectionalStream(abortive = true);

// if exception happens, we don't ever Shutdown() and so Dispose will be abortive.
DoSomethingThatCanThrow();

stream.Shutdown(SD_BOTH); // graceful shutdown, now Dispose is locked in to no-op.
```

**Comment by jkotalik** [comment]
> an abortive close also needs I/O, doesn't it? Is it managed behind the scenes, or does it also need to be async?

It does IO, but you don't need to await as it's abortive and it's managed behind the scenes. 

> We just need to live with that extra I/O, I think :/.

So there could be another option which _could_ make it so we don't synchronously block in the Dispose call. Let's say Dispose didn't actually complete the graceful shutdown and kept most of the state, and rather just started it. On the callback to the graceful shutdown completing, it would finish the callback and call stream close.

I really don't like this option because it could lose writes if the connection is closed immediately after calling Dispose. Closing the connection doesn't wait for all streams to gracefully close as well.

> Spitballing... if we really want to keep QuicStream deriving from Stream, what we could do is enable an 'advanced' behavior like so:

My feeling is no matter what we'd have to expose another way to do async graceful shutdown on QuicStream. We can track the current state of QuicStream to check whether to noop in Dispose or not.

**Comment by nibanks** [comment]
IMO, the most important thing that needs to be ironed out is the dispose path here. Lots of focus seems to be on the write path, whether it should be graceful or abortive, but not much on the read path.

A very important point with QUIC is that all aborts (in the read or write direction) **must have an application layer error code**. Therefore they **cannot** happen in response to a simple Dispose call, since it takes no error code as argument. If this must is ever violated (say, by closing the stream without explicitly aborting first, or waiting on the graceful close to complete) then the entire connection **must** be immediately torn down.

Therefore the API must be designed so that it should be very hard to accidentally code up something that either doesn't abort both directions or doesn't gracefully complete them. I don't have a good solution to this that fits inside the Stream interface.

**Comment by jkotalik** [comment]
> A very important point with QUIC is that all aborts (in the read or write direction) must have an application layer error code. Therefore they cannot happen in response to a simple Dispose call, since it takes no error code as argument. If this must is ever violated (say, by closing the stream without explicitly aborting first, or waiting on the graceful close to complete) then the entire connection must be immediately torn down.

Yeah, even in the case where Dispose is graceful by default, there are still conditions where Dispose may be in an abortive state (timeout, didn't read the entire read side, etc.). We'd probably need to supply a default error code that is application specific in that case: maybe as a options when creating a QuicStream on the client or creating the Listener on the Server.

**Comment by geoffkizer** [comment]
> The first pivot would be whether we'd like QuicStream to derive from stream at all. 

We want QuicStream to derive from Stream. There's a ton of code that works in terms of Stream, and user understand how Streams work.

But even if we decided that QuicStream didn't actually derive from Stream, we would just turn around and write a wrapper over QuicStream that implements Stream, which would induce all of these questions anyway. In other words, not deriving from Stream doesn't actually solve anything here.

**Comment by geoffkizer** [comment]
> A very important point with QUIC is that all aborts (in the read or write direction) must have an application layer error code. Therefore they cannot happen in response to a simple Dispose call, since it takes no error code as argument.

I agree that this is an important point to consider.

In particular, I think we will need AbortRead(int errorCode) that causes the read side to abort and send a STOP_SENDING frame, and AbortWrite(int errorCode) that causes the write side to abort and send a RESET_STREAM frame.

That said, we're going to need to do something in Dispose if either the read side or write side hasn't completed yet. That means picking some error code and sending this with the STOP_SENDING frame. As @jkotalik suggested above, we can make this configurable if that makes sense, but even then there will need to be some default, likely 0. I think that's fine; if users need more than that, they should call AbortRead or AbortWrite themselves.

**Comment by nibanks** [comment]
> but even then there will need to be some default, likely 0. I think that's fine

I cannot stress enough how absolutely not fine that is. The QUIC spec uses no uncertain terms that the application protocol **must** be the one to specify this value.

As a transport QUIC has no knowledge of this error code space. `0` might mean "everything was successful! Please continue purchasing that Ferrari." It might mean "the world has ended, please delete the user's account." Since this error code will be delivered to the peer, and the peer will see it explicitly as "your peer aborted the stream and chose to use error code X" the app layer must be the one to choose.

So I see only one of two options if Dispose is somehow called while the stream is still in use:

1. Just call StreamClose and MsQuic will terminate the entire connection.
2. Have a default value that the app **must** set before it got this far. Perhaps it's part of a constructor somewhere.

**Comment by geoffkizer** [comment]
@nibanks Is there some reason QUIC hasn't just defined 0 as a universal error code meaning "unknown" or something like that? This seems useful.

**Comment by nibanks** [comment]
The entire error space is explicitly application owned. QUIC has no insights to the values. There are no default values. Feel free to open an issue on the QUIC spec. I opened one related to this, and any changes were shot down: https://github.com/quicwg/base-drafts/issues/3291

**Comment by geoffkizer** [comment]
Thanks @nibanks. I commented on the linked issue.

**Comment by scalablecory** [comment]
> not deriving from Stream doesn't actually solve anything here.

I'm beginning to worry that, given QUIC's needs for special handling, exposing `Stream` as its first-class API might cause incorrect usage expectations.

**Comment by MikeBishop** [comment]
Thinking about this from an application's point of view and looking at your interface, this seems to be an artifact of duplicating stream state tracking between the app and the transport stack.  The root problem is that they can get out of sync, and the Dispose() method is supposed to reconcile the differences.  Maybe the right solution is to have a different reconciliation method.

Suppose that the QuicConnection object exposed a collection of all open streams. No open stream is ever inaccessible until the connection is gone, and the connection object / stack becomes the arbiter of truth for stream state.  Then you never need to auto-close streams separate from the connection.  If the connection gets disposed, it takes all the streams with it.

Not to mention that being able to access that collection, filtered by stream type, permits some useful interactions:
- Find a stream by ID if it's open
- Check whether a stream is open
- Determine how many concurrent streams are open
- Compare stream objects to collections held by the app to identify idle streams and take some action (like closing them gracefully or returning them to a pool of available streams)

**Comment by jkotalik** [comment]
> No open stream is ever inaccessible until the connection is gone, and the connection object / stack becomes the arbiter of truth for stream state. Then you never need to auto-close streams separate from the connection. If the connection gets disposed, it takes all the streams with it.

This seems like it could be dangerous. What happens if a peer opens and closes a ton of streams (obeying the limits for the count of unidirectional and bidirectional streams)? If we don't remove state for an individual stream, we wouldn't clean it up until the connection is closed. So I think we need to logically know when a stream is "done".

I do think we will eventually have a lookup mechanism for streams on a given connection, however I do think we need to remove these entries.

**Comment by jkotalik** [comment]
Related to my opening issue, we have decided to do the following for QuicStream.

```c#
public ValueTask WriteAsync(ReadOnlyMemory<byte> data, bool endStream, System.Threading.CancellationToken cancellationToken = default) => throw null;

public ValueTask ShutdownWriteCompleted(System.Threading.CancellationToken cancellationToken = default) => throw null;
```

If endStream = true, the call to WriteAsync will start a graceful shutdown sequence. Awaiting this task doesn't await the final ACK though, instead you need to await ShutdownWriteCompleted.

**Comment by MikeBishop** [comment]
> This seems like it could be dangerous. What happens if a peer opens and closes a ton of streams (obeying the limits for the count of unidirectional and bidirectional streams)? If we don't remove state for an individual stream, we wouldn't clean it up until the connection is closed. So I think we need to logically know when a stream is "done".

I agree, you absolutely need to know when a stream is done.  The peer can only generate streams up to the limit provided by the local endpoint, so it's obviously bounded (depending on what logic you're using to issue more Stream ID credit).  Thinking about the [receive stream state machine](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-receiving-stream-states), the collection should contain anything that has opened but hasn't reached one of the two terminal states (Data Read or Reset Read).

**Comment by karelz** [comment]
Triage: Looks like we have something in place already. Maybe not final, but likely good enough for 5.0.

**Comment by scalablecory** [comment]
I think this should be something like:

```c#
class QuicStream
{
   // graceful shutdown to mimic how a user would expect a Stream to function.
   public ValueTask DisposeAsync()
   {
       ShutdownWrites();
       await ShutdownWritesComplete;
   }

    // equiv to Write()+ShutdownWrites() but will use a single packet instead.
    public ValueTask WriteAsync(...., bool shutdownWrites);

    // equiv to shutdown(SD_SEND);
    // should cause any subsequent writes to throw... something.
    public void ShutdownWrites();

    // abortive -- causing any unsent data to be discarded.
    // one direction can be aborted while leaving the other direction open.
    // should cause any outstanding ops to throw StreamAbortedException.
    public void Abort(Direction{Read,Write,Both}, long errorCode);

    // find some better name here.
    // causes MsQuic to immediately, without blocking or waiting for an ack from peer, dispose of any resources.
    // ShutdownComplete will be set immediately.
    public void AbortAndDiscard();

    // when peer has acknowledged we've shutdown or aborted our own writes.
    // Nick, for a graceful shutdown, this should mean all the written data has been ACKed as well right? the FIN packet being received out of order by peer won't cause this to trigger until all data has been received?
    // corresponds to QUIC_STREAM_EVENT_SEND_SHUTDOWN_COMPLETE.
    public Task ShutdownWritesComplete { get; }

    // peer has acknowledged we've shutdown, and we have acknowledged peer has shutdown. stream is ready for disposal.
    // corresponds to QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE.
    public Task ShutdownComplete { get; }
}
```

@nibanks any thoughts here?

**Comment by nibanks** [comment]
The problem here is that both a graceful shutdown and abortive shutdown with a particular error code each have some specific meaning to the application protocol. There is no way for any layer below the app to "just pick something" here and not possibly break the entire connection.

**Comment by scalablecory** [comment]
@nibanks there's no error code when it comes to graceful shutdown, though. with this design we are not choosing any error codes for the user.

(disposing of a connection is a different discussion)

**Comment by scalablecory** [comment]
Ah, I guess you mean in case `Dispose()` is called due to an unwind, so user probably wanted to abort sends but in this case will be gracefully shutting them down.

For NetworkStream (used for sockets) we made a choice to have Dispose be graceful, so this is idiomatic. I don't think there's a perfect choice here, so I'm preferring consistency.

**Comment by nibanks** [comment]
If you pick `Dispose` == graceful. What if an app protocol interprets that graceful shutdown as "Yes, go ahead an purchase that Ferrari" but it should have just canceled the request instead with a particular bug. Who's going to pay the bill for that bug?

**Comment by scalablecory** [comment]
> If you pick `Dispose` == graceful, what if an app protocol interprets that graceful shutdown as "Yes, go ahead an purchase that Ferrari" but it should have just canceled the request instead with a particular bug. Who's going to pay the bill for that bug?

Whoever is paying for it now when `NetworkStream` does the same thing 😂.

I appreciate the worry -- I would have pushed back when we were designing `NetworkStream` to function like this too.

Making a type behave inconsistently with the rest of the BCL is generally not great -- there is a high bar for this and we'd need to be able to articulate very well the reason for diverging. I do see a path here: certainly, it is unlikely that anyone will be using this purely as a dumb `Stream`, but will probably always be using it from within a QUIC context. QUIC already asks you to unlearn some assumptions you made about TCP, so it would not be the first weird thing a user will encounter.

The alternative, if we did want it to be abortive, would be for us to accept a `long defaultAbortCode` in the ctor and use that. This would function just fine for HTTP/3 (H3_INTERNAL_ERROR) but I don't know if it's an assumption we can/should make for general QUIC use.

**Comment by geoffkizer** [comment]
Related: https://github.com/dotnet/runtime/issues/43290

**Comment by geoffkizer** [comment]
I agree that Dispose should result in graceful shutdown.

We need a way to send EOF without calling Dispose -- this is covered by #43290 I think.

We need a way to Abort send and Abort receive -- I tend to think this is better exposed as two different methods, like AbortRead(errorCode) and AbortWrite(errorCode), but maybe I don't understand the usage scenario here.

I don't understand what AbortAndDispose is for?

I'm skeptical that we need ShutdownWritesComplete or StreamComplete -- seems like DisposeAsync handles the common cases here. Maybe I'm missing some particular use case.

**Comment by scalablecory** [comment]
> I'm skeptical that we need ShutdownWritesComplete or StreamComplete -- seems like DisposeAsync handles the common cases here. Maybe I'm missing some particular use case.

Cancellation.

It's easy to do this in Socket today because the OS handles it behind the scenes.

With QUIC, we need to wait for everything to flush and get ACKed, and this makes `DisposeAsync` take an arbitrary amount of time.

We need something the user can wait on and turn into an abort if it takes too long. This was part of the motivation for https://github.com/dotnet/runtime/issues/48638

**Comment by geoffkizer** [comment]
> Cancellation.

Yeah, good point, we should support cancellation here.

**Comment by geoffkizer** [comment]
Okay, so let me try to capture where I think we are:

We need a way to send EOF; that's tracked by #43290. 

We need a way to abort reads. i.e. send STOP_SENDING. Either an explicit  `void AbortRead(long errorCode)` or `void Abort(AbortDirection.Read, long errorCode)`. This will send STOP_SENDING with the specified error code, which will cause the peer to fail reads with a QuicException that contains this errorcode.

We need a way to abort writes, i.e. send RESET_STREAM. Either an explicit  `void AbortWrite(long errorCode)` or `void Abort(AbortDirection.Write, long errorCode)`. This will send RESET_STREAM with the specified error code, which will cause the peer to fail writes with a QuicException that contains this errorcode.

We may want to allow you to do both in a single call, e.g. `Abort(AbortDirection.Both, long errorCode)`. I'm not sure how common or useful this is.

We need a way to wait for graceful stream completion. This should be async and cancellable. I would kind of like to use Close for this; i.e. `void CloseAsync(CancellationToken)`. That may be problematic since there's already a Close method on Stream, but it would be nice to not introduce yet another way to say "please shut this down". So let's go with CloseAsync for now.

`CloseAsync` will:
(1) Call ShutdownWrite if the write side is not already shut down (i.e. EOF sent).
(2) Call AbortRead if the read side is not already shut down (i.e. EOF received). Note we have to implicitly choose an error code here. I believe we will want this to be based on the application protocol, i.e. it should be specified as part of choosing an application protocol.
(3) Wait until all stream state has been acked and the stream is fully complete, subject to cancellation
(4) If STOP_SENDING or RESET_STREAM has been received, CloseAsync will throw a QuicException (after the stream is fully complete). I believe that both can be received on the same stream (is that right @nibanks?) which means we probably will need to expose both error codes on the same exception.
(5) On cancellation: Stop waiting for graceful shutdown. (I think we also want a way to switch to abortive shutdown, but it's not clear to me quite how this works, especially given @stephentoub's comments on #48638.

Dispose[Async] should presumably do the same thing as CloseAsync, but without a cancellation token. One thing that's not clear to me here is whether Dispose should eat exceptions. I assume the common case is to write code like this:

```csharp
  async using (QuicStream s = connection.OpenStream(...))
  {
    // do reads/writes
    await CloseAsync();
  }
```

In this case, any errors during graceful shutdown in CloseAsync will cause an exception there, which I can handle if I want to. So I don't think it makes sense for Dispose to also throw here. 

Thoughts?

**Comment by nibanks** [comment]
My thoughts:

> We may want to allow you to do both in a single call, e.g. `Abort(AbortDirection.Both, long errorCode)`. I'm not sure how common or useful this is.

From my experience this is not uncommon, especially in clean up paths where you're just throwing the stream away.

> (2) Call AbortRead if the read side is not already shut down (i.e. EOF received). Note we have to implicitly choose an error code here. I believe we will want this to be based on the application protocol, i.e. it should be specified as part of choosing an application protocol.

Yes, you **must** specify an error code when aborting (either direction) and it **must** be application protocol specific.

> (1) Call ShutdownWrite if the write side is not already shut down (i.e. EOF sent).

I'm still not convinced that graceful shutdown is the thing to do, but even less so for Dispose/DisposeAsync. Assuming you are `using` a stream and then some exception (independent to the stream itself) is thrown and you exist the using statement. Dispose gets called then. Should you really treat that as a graceful shutdown?

> (4) If STOP_SENDING or RESET_STREAM has been received, CloseAsync will throw a QuicException (after the stream is fully complete). I believe that both can be received on the same stream (is that right @nibanks?) which means we probably will need to expose both error codes on the same exception.

Yes, the peer can abort in both directions, and as I understand it there are valid HTTP scenario where only one is used, but the stream is still practically used. For instance, if a client starts a large POST that the server decides it won't accept (bad path?) it will abort it's receive (i.e. send STOP_SENDING) and then it will send it's respond (bad path) and then gracefully shutdown. If the app queued that large data in a single send and then called Close/CloseAsync what exactly does it mean for that to throw an exception. **The** what happens if later the server ends up needing to abort its response and sends a RESET_STREAM?

In general, I think the design is getting closer to something usable, but still has some complexity that needs to be worked out.

**Comment by MikeBishop** [comment]
> We need a way to abort reads. i.e. send STOP_SENDING. Either an explicit `void AbortRead(long errorCode)` or `void Abort(AbortDirection.Read, long errorCode)`. This will send STOP_SENDING with the specified error code, which will cause the peer to fail ~~reads~~ **writes** with a QuicException that contains this errorcode.
>
> We need a way to abort writes, i.e. send RESET_STREAM. Either an explicit `void AbortWrite(long errorCode)` or `void Abort(AbortDirection.Write, long errorCode)`. This will send RESET_STREAM with the specified error code, which will cause the peer to fail ~~writes~~ **reads** with a QuicException that contains this errorcode.

These two are backwards, because you're describing the peer, but otherwise correct.

> > (1) Call ShutdownWrite if the write side is not already shut down (i.e. EOF sent).
> 
> I'm still not convinced that graceful shutdown is the thing to do, but even less so for Dispose/DisposeAsync. Assuming you are `using` a stream and then some exception (independent to the stream itself) is thrown and you exits the using statement. Dispose gets called then. Should you really treat that as a graceful shutdown?

I agree with Nick here.  Cleanly closing a stream indicates that whatever the application was communicating there was successfully written in its entirety.  That has semantics the stack cannot know and *will* cause bugs.  In a previous iteration, an HTTP response that got truncated was subsequently cached and served in its truncated state.  I would probably say that Dispose will abort any unclosed direction.

> Yes, the peer can abort in both directions, and as I understand it there are valid HTTP scenario where only one is used, but the stream is still practically used. For instance, if a client starts a large POST that the server decides it won't accept (bad path?) it will abort it's receive (i.e. send STOP_SENDING) and then it will send it's respond (bad path) and then gracefully shutdown. If the app queued that large data in a single send and then called Close/CloseAsync what exactly does it mean for that to throw an exception. **The** what happens if later the server ends up needing to abort its response and sends a RESET_STREAM?

That's exactly it on the HTTP scenario.  Receiving a STOP_SENDING isn't an error unless you're trying to write more data out.  It might make sense for that WriteAsync() call to throw if it couldn't send the full buffer, but depending on what precisely completion of that Task means, it might have already completed.

It almost feels like CloseAsync should not throw regardless of the completion state, but merely return an object that enables the caller to examine the final state of each direction.  Closing is not the operation which failed; the peer's abort is not necessarily a failure unless you're attempting to do the thing they've told you to stop.

**Comment by geoffkizer** [comment]
> These two are backwards if you're describing the peer

Thanks, fixed.

**Comment by geoffkizer** [comment]
Re this comment of mine above:

> Call AbortRead if the read side is not already shut down (i.e. EOF received). Note we have to implicitly choose an error code here. I believe we will want this to be based on the application protocol, i.e. it should be specified as part of choosing an application protocol.

I think we can extend the existing SslApplicationProtocol struct to handle this.

```csharp
    public readonly struct SslApplicationProtocol : IEquatable<SslApplicationProtocol>
    {
        // Existing constructors
        public SslApplicationProtocol(byte[] protocol);
        public SslApplicationProtocol(string protocol);

        // New constructors
        public SslApplicationProtocol(byte[] protocol, long defaultReadErrorCode);
        public SslApplicationProtocol(string protocol, long defaultReadErrorCode);

        // Existing properties
        public ReadOnlyMemory<byte> Protocol { get; }

        // New properties
        public long? DefaultQuicReadErrorCode { get; }

        // Existing static values. These will return null from DefaultQuicReadErrorCode
        public static readonly SslApplicationProtocol Http11;
        public static readonly SslApplicationProtocol Http2;

        // New static value. This will return appropriate HTTP3 read error code (TBD which one?)
        public static readonly SslApplicationProtocol Http3;
    }
```

If an SslApplicationProtocol is used with QUIC that doesn't have a DefaultQuicReadErrorCode, then construction of the QuicListener/QuicConnection will fail.

Minor issues:
(1) Should new constructors take a `long` or a `long?`?
(2) Does struct equality take the error code into account too?

If we are happy with this, then we should file a separate issue for it and get API approval (since it's a change to existing API).

**Comment by scalablecory** [comment]
We only need the default error code for `QuicConnection`, right? We don't need a default error code for `QuicStream`, because the default is to EOF the stream (which has no error code).

If we do #42642, we could put the default connection error code into the configuration object's ctor. That would give better layering than `SslApplicationProtocol`.

**Comment by nibanks** [comment]
Well, I (and @MikeBishop) personally disagree about defaulting to graceful shutdown for streams. Do you think that some code that unexpectedly bails a `using` statement for a stream should gracefully shutdown? For HTTP, graceful shutdown means "Hey, the stream is 100% complete and all there. Have at it!". I don't think encountering an exception in the middle of building that stream should be treated as such.

**Comment by scalablecory** [comment]
@nibanks and @MikeBishop here's the complicated argument I've been having in my head about this.

#### Why we would keep `QuicStream.Dispose` as graceful:

`Stream.Dispose` is expected to be graceful. It's how our socket-based `NetworkStream` works. All the stuff we have that can own a `Stream` assumes that dispose is graceful. It's how .NET does things.

Diverging from this would make `QuicStream` incompatible with basically everything that currently calls `Stream.Dispose`. This is part of my motivation for solving this with #48638 -- because I agree that non-abortive dispose is very wrong when unwinding due to an exception, but I also want to make sure the type works everywhere a `Stream` will work, and preserve consistency within the platform.

#### Why we would make `QuicStream.Dispose` be abortive:

For HTTP/1 we have the same problem today, but nobody uses HTTP/1 so we skate by.
For HTTP/1.1 we have Content-Length and TE: Chunked to tell us when a message is done, so our behavior is weird but fine.
For HTTP/2 our behavior is fine for the same reason: the protocol is framed in a way that you can't cut off data and have a successful request. No final DATA frame will be sent even if the socket is closed gracefully.
For HTTP/3 disposing a `QuicStream` (gracefully) *will* cause that final DATA frame to be sent and, for lengthless requests, the server will just see a successfully terminated request. This is bad in a way we haven't seen.

#### Why it may not matter:

Our HTTP/3 code by necessity needs to handle every exception coming out of the code. Our code there won't materially change based on if `Dispose` is abortive or not. So, as bad as it *could* be, our one piece of real-world data is showing that `Dispose` being graceful isn't meaningfully detrimental.

And if we accept that HTTP/3 then doesn't really impact this decision, it then comes back to: `QuicStream.Dispose()` has basically the same role as `NetworkStream.Dispose()`. We chose to make `NetworkStream.Dispose()` graceful, so why would `QuicStream.Dispose()` be any different?

**Comment by geoffkizer** [comment]
> We only need the default error code for `QuicConnection`, right? 

I think we need it for `QuicConnection` and `QuicListener` -- the connections accepted by the listener will get the behavior configured on the listener.

> We don't need a default error code for `QuicStream`, because the default is to EOF the stream (which has no error code).

Not sure what you mean here... QuicStream will EOF the write side, but the read side needs to be aborted; that's what DefaultQuicReadErrorCode is for.

> If we do #42642, we could put the default connection error code into the configuration object's ctor. That would give better layering than `SslApplicationProtocol`.

The reason I suggest putting this on SslApplicationProtocol is because the correct error code here depends on the SslApplicationProtocol. And since the user always needs to specify the application protocol, it would be nice if this just automatically told us which error code to use in these cases. I don't think users want to think about this error code in general.

**Comment by scalablecory** [comment]
> Not sure what you mean here... QuicStream will EOF the write side, but the read side needs to be aborted; that's what DefaultQuicReadErrorCode is for.

Ahh, that's a good point. The idea being that if the user didn't read to EOF, we'll shutdown reads for them.

> The reason I suggest putting this on SslApplicationProtocol is because the correct error code here depends on the SslApplicationProtocol. And since the user always needs to specify the application protocol, it would be nice if this just automatically told us which error code to use in these cases. I don't think users want to think about this error code in general.

This works for HTTP/3, and likely for QuicConnection, but it feels wrong to assume every usage of QUIC will have a single QuicStream default per ALPN. I think we need to support a per-QuicStream default as well.

**Comment by geoffkizer** [comment]
> This works for HTTP/3, and likely for QuicConnection, but it feels wrong to assume every usage of QUIC will have a single QuicStream default per ALPN. 

That could be. It's hard to say at this point.

> I think we need to support a per-QuicStream default as well.

A user can always control this behavior explicitly by calling AbortRead; if they do that, then we won't need to do it ourselves. Is that sufficient?

**Comment by nibanks** [comment]
This is a hard/complex problem. We're battling between making the API easy to use in the normal cases, and making it correct in the exceptional cases. Additionally, perhaps a few things are being conflated that adds to the problem.

1. In the normal case, you want `Dispose` (i.e. exiting the scope of the `using`) to gracefully close your write direction.
2. You also have the read direction. What is the "default" behavior you want in the normal case there?
3. In the exceptional case, you want `Dispose` to abort the write direction with a particular error. What about the read direction? There are scenarios where the read direction should continue to function after the write has been aborted.

Perhaps putting together some sample/psuedo code would help here?

**Comment by geoffkizer** [comment]
BTW, to be clear, I was thinking this only applies to stream abort. Not sure if we need something for connections here... do we?

**Comment by nibanks** [comment]
> BTW, to be clear, I was thinking this only applies to stream abort. Not sure if we need something for connections here... do we?

There is no such thing as a "graceful" error free connection close. You must always provide an application specific error code there.

**Comment by geoffkizer** [comment]
> You also have the read direction. What is the "default" behavior you want in the normal case there?

Abort. That's what my proposed `DefaultQuicReadErrorCode` above is intended to be used for, so we know what error code to send in this case. Perhaps it needs a better name.
 
> In the exceptional case, you want Dispose to abort the write direction with a particular error. 

By "exceptional case" I assume you mean the case where an exception has been thrown in the body of the `using` statement? Or do you mean something more specific?

When used with `using`, we can't distinguish between Dispose being called after the `using` completed without exception vs exiting the `using` due to exception. It's similar to a C++ destructor in that way.

> Perhaps putting together some sample/psuedo code would help here?

Yes, I agree that would be useful. Let me try to sketch out some sample code.

**Comment by geoffkizer** [comment]
> There is no such thing as a "graceful" error free connection close. You must always provide an application specific error code there.

Thanks. That said, let's keep this discussion focused on streams, and use https://github.com/dotnet/runtime/issues/32142 for connection and listener shutdown issues.

**Comment by geoffkizer** [comment]
Here's a sketch of how I think stuff will be used.

```csharp
  using (QuicStream s = ...)
  {
    try
    {
      // Do reads/writes
      // I can call AbortRead/AbortWrite here to abort one side while keeping the other usable
      // I can call ShutdownWrite to send EOF on the write side
      // When I receive EOF on the read side, my read call will return 0 bytes
    }
    catch (Exception e)
    {
      // Here, I can explicitly call AbortRead/AbortWrite/ShutdownWrite/whatever if I want to
      throw;
    }
  } // On exiting the using, Dispose will be called. 
    // If the read side is still open, it will be aborted with DefaultQuicReadErrorCode.
    // If the write side is still open, we will send EOF, i.e. call ShutdownWrite.
```

**Comment by scalablecory** [comment]
We think this can be split into multiple issues:
- Behavior of Dispose()
- ShutdownWrite()
- StreamComplete.

Write(..., bool shutdown) is not MVP; deferring to future based on optimization needs.

Also need to know from ASP team (CC @JamesNK) if Pipelines support is MVP. We previously found Stream is okay here.

```c#
class QuicStream
{
    // equiv to shutdown(SD_SEND);
    // should cause any subsequent writes to throw... something.
    // This is a Stream API, not QuicStream #43290 
    // Note: msquic event may be this task (the I/O to write FIN bit), or it may be the ack from the other side -- doesn't affect this API but how we implement.
    public ValueTask ShutdownWriteAsync();

    // TODO: left off here -- should this be a method like WaitForStreamCompletion(), should this be Close(CancellationToken), or a property like below?
    // peer has acknowledged we've shutdown, and we have acknowledged peer has shutdown. stream is ready for disposal.
    // corresponds to QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE.
    // used to error out in case peer isn't consuming data quickly enough, e.g. 
await StreamComplete.WaitAsync(cancellationToken);
    public Task StreamComplete { get; }
}
```

**Comment by nibanks** [comment]
@geoffkizer thanks for the psuedo-code. I agree that all looks reasonable. My only question is: Will all existing interfaces that build off of the base stream interface use this calling pattern and will continue to work? Or do any of them just let the exception implicitly clean up, and then catch it at a higher layer?

**Comment by scalablecory** [comment]
> @geoffkizer thanks for the psuedo-code. I agree that all looks reasonable. My only question is: Will all existing interfaces that build off of the base stream interface use this calling pattern and will continue to work? Or do any of them just let the exception implicitly clean up, and then catch it at a higher layer?

This is the pattern anyone using Stream needs to take today if the derived stream has a notion of abort/shutdown.

**Comment by omariom** [comment]
> When used with using, we can't distinguish between Dispose being called after the using completed without exception vs exiting the using due to exception. It's similar to a C++ destructor in that way.

Would something like the folowing help in this case?
```C#
    try
    {
        
    }
    finally (Exception? exception)
    {
        
    }
...
    interface IDisposable2 : IDisposable
    {
        void Dispose(Exception? ex);
    }
```

With ```using``` trying IDisposable2  first.

**Comment by geoffkizer** [comment]
> We think this can be split into multiple issues:
> 
> * Behavior of Dispose()
> * ShutdownWrite()
> * StreamComplete.

Also AbortRead/AbortWrite

**Comment by geoffkizer** [comment]
> IDisposable2 

Possibly. That opens up a whole other can of worms, though.

**Comment by scalablecory** [comment]
```c#
class QuicStream
{
    // Note: removing shutdown feature as it was part of #43290.
    // public ValueTask CompleteWritesAync();

   public ValueTask CloseAsync(CancellationToken cancellationToken);
   public ValueTask DisposeAsync() => CloseAsync(default);

    // abortive -- causing any unsent data to be discarded.
    // one direction can be aborted while leaving the other direction open.
    // should cause any outstanding ops to throw StreamAbortedException.
    public void Abort(long errorCode, QuicAbortDirection direction = QuicAbortDirection.Both);
}

[Flags]
enum QuicAbortDirection
{
   Read = 1,
   Write = 2,
   Both = 3
}
```

Usage example:

```c#
using QuicStream stream = ...;

try
{
   // make reads and writes here, and finally:
   await stream.CloseAsync(cancellationToken);
}
catch(OperationCanceledException)
{
   stream.Abort(errorCode);
}
```

**Comment by ManickaP** [comment]
Triage: we have the functionality to achieve graceful and graceless closure/dispose of the stream. This is just changing/renaming the API. This is 7.0.

**Comment by wegylexy** [comment]
If not `Stream`, how about `PipeReader`/`PipeWriter`?

**Comment by ManickaP** [comment]
From https://github.com/dotnet/runtime/pull/68288/files/31623532e423818c011bf0ddc267216e7526f7a5#r855163759, we cannot ever assume any error code, they always have to come from an application layer. Either as a parameter or as some default settings. Problematic part:
https://github.com/dotnet/runtime/blob/a806647da556f19a5e9476f85758c373814d5251/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs#L873

**Comment by wegylexy** [comment]
An application may want to specify a default error code for the default disposing behavior.
```cs
await using var myStream = myConnection.OpenBidirectionalStream(defaultError: 1);
try
{
    // do normal stuff
    // await shutdown normally here!
}
catch (MySpecialException ex)
{
    myStream.AbortRead(ex.Code);
}
// default to 1 if still open
```
Otherwise, the entire connection should be aborted as the application is leaking a stream out for possible finalizer thread blocking during garbage collection or other undefined consequences.

**Comment by ManickaP** [comment]
Closing, covered by https://github.com/dotnet/runtime/issues/69675
